fingerprint/infra/scanner/src/main/java/com/simprints/fingerprint/infra/scanner/v2/channel/MainMessageChannel.kt

```diff
@@ -6,13 +6,31 @@ import com.simprints.fingerprint.infra.scanner.v2.incoming.main.MainMessageInput
 import com.simprints.fingerprint.infra.scanner.v2.outgoing.main.MainMessageOutputStream
 import com.simprints.fingerprint.infra.scanner.v2.tools.reactive.doSimultaneously
 import io.reactivex.Single
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.rx2.rxSingle
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.rx2.await
 
 class MainMessageChannel(
-    incoming: MainMessageInputStream,
-    outgoing: MainMessageOutputStream
+    incoming: MainMessageInputStream, outgoing: MainMessageOutputStream
 ) : MessageChannel<MainMessageInputStream, MainMessageOutputStream>(incoming, outgoing) {
 
-    inline fun <reified R : IncomingMainMessage> sendMainModeCommandAndReceiveResponse(command: OutgoingMainMessage): Single<R> =
-        outgoing.sendMessage(command)
-            .doSimultaneously(incoming.receiveResponse())
+    // Mutex to ensure that only one thread can send and receive messages at a time
+    val mutex = Mutex()
```

Comment:
Should the same apply to other channels as well?

Reply:
Yes, you're right; we should use the same logic for all channels. But the problem is that testing the OTA and other scanner flows will be hard, and there haven’t been any reports of concurrency issues with the other channels.

Reply:
Maybe worth making a ticket to remember to look into it for upcoming major releases.